test/model/compose_test.dart

```diff
@@ -0,0 +1,190 @@
+import 'package:checks/checks.dart';
+import 'package:test/scaffolding.dart';
+import 'package:zulip/model/compose.dart';
+
+void main() {
+  group('wrapWithBacktickFence', () {
+    /// Check `wrapWithBacktickFence` on example input and expected output.
+    ///
+    /// The intended input (content passed to `wrapWithBacktickFence`)
+    /// is straightforward to infer from `expected`.
+    /// To do that, this helper takes `expected`, validates it
+    /// (as a courtesy for the test author), and removes the opening and
+    /// closing fences.
+    ///
+    /// Then we have the input to the test, as well as the expected output.
+    void checkFenceWrap(String expected, {String? infoString}) {
+      final lines = expected.split('\n');
+      final firstLineMatch = RegExp(r'^(`{3,})[^`]*$').allMatches(lines.removeAt(0)).singleOrNull;
+      assert(firstLineMatch != null,
+        'test error: opening fence not found in `expected`');
+      assert(lines.removeAt(lines.length - 1) == '',
+        'test error: `expected` should end with a newline');
+      final openingFenceLength = firstLineMatch![1]!.length;
+      assert(RegExp(r'^`{' + openingFenceLength.toString() + r'}$').hasMatch(lines.removeAt(lines.length - 1)),
+        'test error: closing fence not found in `expected`');
+      final content = lines.join('\n');
```

Comment:
Here's another way to express the same set of allowed `expected` and resulting `content`:
```suggestion
      final re = RegExp(r'^(`{3,})[^`]*?\n(.*)\n\1\n$', dotAll: true);
      final content = re.firstMatch(expected)![2]!;
```

I think it's probably actually better, though, to simplify away this helper's knowledge of things like backticks and matching fence lengths, and let the code under test speak for itself in that respect. That then looks like this:
```dart
      final re = RegExp(r'^.*?\n(.*)\n.*\n$', dotAll: true);
      final content = re.firstMatch(expected)![1]!;
```
and the helper's doc can just say it removes the first and last line to get the input.

Then if the test passes something with, say, mismatched fences:
```dart
      checkFenceWrap('''
````
hello world
```
''');
```

the developer gets output from `check` like:
```
00:01 +0 -1: wrapWithBacktickFence single line with no code blocks [E]         
  Expected: a String that:
    equals '````
    hello world
    ```'
  Actual: '```
  hello world
  ```'
  Which: differs at offset 3:
  ````\nhello w ...
  ```\nhello wo ...
     ^
  package:checks/src/checks.dart 72:9               check.<fn>
  package:checks/src/checks.dart 717:12             _TestContext.expect
  package:checks/src/extensions/string.dart 108:13  StringChecks.equals
  test/model/compose_test.dart 20:78                main.<fn>.checkFenceWrap
  test/model/compose_test.dart 24:7                 main.<fn>.<fn>
```
which seems pretty clear, and appropriate — the result of the test is that the code under test did not produce the output that the test said was expected.

Reply:
In order to make `content` consist of complete lines as discussed in my other comment, the capturing group just needs to enclose the newline:
```dart
      final re = RegExp(r'^.*?\n(.*\n).*\n$', dotAll: true);
```

Then in order to test that the function correctly handles the case of an incomplete final line too, we can add another flag to the helper:
```dart
    void checkFenceWrap(String expected, {String? infoString, bool chopNewline = false}) {
      final re = RegExp(r'^.*?\n(.*\n).*\n$', dotAll: true);
      String content = re.firstMatch(expected)![1]!;
      if (chopNewline) content = content.substring(0, content.length - 1);
```

and have a couple of test cases using that.